# Aprendizado de Máquina - Classificação

Nome: Ricardo Coutinho Cordeiro

Professor: Ulisses Brisolara Corrêa


## Baixar dataset

In [ ]:
!pip install gdown -qqq

In [ ]:
!gdown https://drive.google.com/uc?id=16OWdq_ebnlmF7xr5JZc4NzRTRAuthoa5

Downloading...
From: https://drive.google.com/uc?id=16OWdq_ebnlmF7xr5JZc4NzRTRAuthoa5
To: /content/cars-india-dataset.csv
100% 20.6k/20.6k [00:00<00:00, 46.9MB/s]


In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier


## Carregar e visualizar conjunto de dados
Material de apoio:
- https://www.w3schools.com/python/pandas/pandas_csv.asp
- https://pandas.pydata.org/docs/user_guide/10min.html#viewing-data

In [ ]:
df = pd.read_csv('cars-india-dataset.csv')

df.head()

Model       Maker          Type  Seats  Displacement  Length  Width  \
0  Virtus  Volkswagen         Sedan      5         999.0    4561   1752   
1  Virtus  Volkswagen         Sedan      5         999.0    4561   1752   
2  Virtus  Volkswagen         Sedan      5        1498.0    4561   1752   
3  Taigun  Volkswagen  Mid Size SUV      5         999.0    4221   1760   
4  Taigun  Volkswagen  Mid Size SUV      5         999.0    4221   1760   

   Height  Wheelbase  No_of_Cylinders  ... Rear Brake Drive Turning Radius  \
0    1507       2651              3.0  ...       Drum   2WD           5.05   
1    1507       2651              3.0  ...       Drum   2WD           5.05   
2    1507       2651              4.0  ...       Drum   2WD           5.05   
3    1612       2651              3.0  ...       Drum   2WD           5.05   
4    1612       2651              3.0  ...       Drum   2WD           5.05   

  Fuel Tank Capacity Boot Space Fuel Efficiency  Emission Type   Tyre Size  \
0               45.0      521.0           19.40          BS VI  205/55 R16   
1               45.0      521.0           18.12          BS VI  205/55 R16   
2               45.0      521.0           18.67          BS VI  205/55 R16   
3               50.0      385.0           19.20          BS VI  205/55 R17   
4               50.0      385.0           17.23          BS VI  205/55 R17   

   Variants  NCAP Rating  
0         3   Not Tested  
1         2   Not Tested  
2         1   Not Tested  
3         3            5  
4         2            5  

[5 rows x 24 columns]

In [ ]:
df.shape

(156, 24)

In [ ]:
df.columns

Index(['Model', 'Maker', 'Type', 'Seats', 'Displacement', 'Length', 'Width',
       'Height', 'Wheelbase', 'No_of_Cylinders', 'Fuel', 'Engine Type',
       'Transmission', 'Front Brake', 'Rear Brake', 'Drive', 'Turning Radius',
       'Fuel Tank Capacity', 'Boot Space', 'Fuel Efficiency', 'Emission Type',
       'Tyre Size', 'Variants', 'NCAP Rating'],
      dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 24 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Model               156 non-null    object 
 1   Maker               156 non-null    object 
 2   Type                156 non-null    object 
 3   Seats               156 non-null    int64  
 4   Displacement        155 non-null    float64
 5   Length              156 non-null    int64  
 6   Width               156 non-null    int64  
 7   Height              156 non-null    int64  
 8   Wheelbase           156 non-null    int64  
 9   No_of_Cylinders     155 non-null    float64
 10  Fuel                156 non-null    object 
 11  Engine Type         156 non-null    object 
 12  Transmission        155 non-null    object 
 13  Front Brake         156 non-null    object 
 14  Rear Brake          156 non-null    object 
 15  Drive               156 non-null    object 
 16  Turning 

In [ ]:
df.isnull().sum()

Model                  0
Maker                  0
Type                   0
Seats                  0
Displacement           1
Length                 0
Width                  0
Height                 0
Wheelbase              0
No_of_Cylinders        1
Fuel                   0
Engine Type            0
Transmission           1
Front Brake            0
Rear Brake             0
Drive                  0
Turning Radius        19
Fuel Tank Capacity     1
Boot Space            15
Fuel Efficiency       23
Emission Type          1
Tyre Size              0
Variants               0
NCAP Rating            0
dtype: int64

In [ ]:
np.unique(df['NCAP Rating'])

array(['0', '2', '3', '4', '5', '6', 'Not Rated', 'Not Tested'],
      dtype=object)

In [ ]:
df.describe()

Seats  Displacement       Length        Width       Height  \
count  156.000000    155.000000   156.000000   156.000000   156.000000   
mean     5.538462   1598.109677  4324.455128  1800.512821  1653.666667   
std      1.018440    504.656969   325.767389    79.050452   125.170559   
min      2.000000    799.000000  3731.000000  1579.000000  1370.000000   
25%      5.000000   1197.000000  3995.000000  1751.500000  1544.000000   
50%      5.000000   1493.000000  4315.000000  1790.000000  1631.000000   
75%      6.000000   1999.000000  4598.000000  1855.000000  1755.000000   
max      9.000000   3799.000000  5115.000000  2099.000000  1995.000000   

         Wheelbase  No_of_Cylinders  Turning Radius  Fuel Tank Capacity  \
count   156.000000       155.000000      137.000000          155.000000   
mean   2633.358974         3.870968        5.247080           48.725806   
std     128.228372         0.389966        0.284181           10.997815   
min    2422.000000         3.000000        4.700000           28.000000   
25%    2500.000000         4.000000        5.100000           40.000000   
50%    2610.000000         4.000000        5.200000           50.000000   
75%    2750.000000         4.000000        5.300000           57.000000   
max    3060.000000         6.000000        5.900000           80.000000   

       Boot Space  Fuel Efficiency    Variants  
count  141.000000       133.000000  156.000000  
mean   377.673759        18.028947    2.871795  
std    105.015951         3.273723    2.864055  
min    180.000000         9.000000    1.000000  
25%    296.000000        16.200000    1.000000  
50%    392.000000        18.100000    2.000000  
75%    447.000000        20.250000    3.000000  
max    625.000000        28.000000   20.000000

# Pré-Processamento de Dados
## Tratamento dos dados
Antes de preencher valores faltantes, divida seus dados em conjuntos de treino e teste. Calcule estatísticas apenas no conjunto de treino e use-as para preencher os valores faltantes em ambos os conjuntos. Isso impede vazamento de informações do teste para o treino.

Materiais de apoio:
- https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
- https://www.freecodecamp.org/news/data-cleaning-and-preprocessing-with-pandasbdvhj/#how-to-handle-missing-values
- https://scikit-learn.org/stable/common_pitfalls.html



In [ ]:
# convertendo unidades de comprimento
def transform_data(df):
    df['Length'] = df['Length'] / 1000
    df['Width'] = df['Width'] / 1000
    df['Height'] = df['Height'] / 1000
    return df

df = transform_data(df)

# calcula massa e volume
df['Volume'] = df['Length'] * df['Width'] * df['Height'] # m³
df['Mass'] = df['Volume'] * 900 # kg


In [ ]:
# remove linhas com valores "not rated" ou "not tested"
df = df[~df['NCAP Rating'].isin(['Not Rated', 'Not Tested'])]

In [ ]:
# conjuntos de treino e teste
train_df, test_df = train_test_split(df, test_size=0.3, random_state=42)

# conversao no conjunto de treino e teste
train_df = transform_data(train_df)

train_df['Volume'] = train_df['Length'] * train_df['Width'] * train_df['Height'] # m³
train_df['Mass'] = train_df['Volume'] * 900 # kg

# remove linhas com valores "not rated" ou "not tested"
train_df = train_df[~train_df['NCAP Rating'].isin(['Not Rated', 'Not Tested'])]

# características e rótulos do conjunto de treino
X_train = train_df['Mass'].values.reshape(-1, 1)
y_train = train_df['NCAP Rating']

# transformações no conjunto de teste
test_df = transform_data(test_df)
test_df['Volume'] = test_df['Length'] * test_df['Width'] * test_df['Height'] # m³
test_df['Mass'] = test_df['Volume'] * 900 # kg
test_df = test_df[~test_df['NCAP Rating'].isin(['Not Rated', 'Not Tested'])]

# características e rótulos do conjunto de teste
X_test = test_df['Mass'].values.reshape(-1, 1)
y_test = test_df['NCAP Rating'].values.reshape(-1, 1)

## Transformação dos dados
Novamente, deve-se tomar cuidado com o vazamento de dados. O último link fornecido aqui demonstra como realizar transformações sem ocorrer vazamento de dados.

Material de apoio:
- https://medium.com/@meritshot/standardization-v-s-normalization-6f93225fbd84#:~:text=Standardization%2C%20interestingly%2C%20refers%20to%20setting,data%20onto%20the%20unit%20sphere.
- https://www.shiksha.com/online-courses/articles/normalization-and-standardization/
- https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html
- https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html
- https://machinelearningmastery.com/data-preparation-without-data-leakage/

## Treinamento dos modelos


### Árvore de Decisão
Material de apoio:
-  https://scikit-learn.org/stable/modules/tree.html

In [ ]:
# árvore de decisão
clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)

# previsões
y_pred = clf.predict(X_test)

# avaliar o modelo
print(classification_report(y_test, y_pred, zero_division=0))
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")

              precision    recall  f1-score   support

           0       0.00      0.00      0.00         2
           2       1.00      1.00      1.00         2
           3       0.86      0.75      0.80         8
           4       0.50      0.75      0.60         8
           5       0.87      0.93      0.90        14
           6       1.00      0.33      0.50         3

    accuracy                           0.76        37
   macro avg       0.70      0.63      0.63        37
weighted avg       0.76      0.76      0.74        37

Accuracy: 0.76


### Máquina de Suporte de Vetores
Material de apoio:
- https://scikit-learn.org/stable/modules/svm.html

In [ ]:
# SVM
svm = SVC(kernel='linear', random_state=42)  # kernel linear para simplificar
svm.fit(X_train, y_train)

# previsões
y_pred = svm.predict(X_test)

# avaliar o modelo
print(classification_report(y_test, y_pred, zero_division=0))
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")

              precision    recall  f1-score   support

           0       0.00      0.00      0.00         2
           2       0.00      0.00      0.00         2
           3       0.22      1.00      0.36         8
           4       0.00      0.00      0.00         8
           5       0.00      0.00      0.00        14
           6       0.00      0.00      0.00         3

    accuracy                           0.22        37
   macro avg       0.04      0.17      0.06        37
weighted avg       0.05      0.22      0.08        37

Accuracy: 0.22


### KNN
Material de apoio:
- https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html

In [ ]:
# normalizar os dados
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# modelo KNN
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train_scaled, y_train)

# previsões
y_pred = knn.predict(X_test_scaled)

# avaliar o modelo
print(classification_report(y_test, y_pred, zero_division=0))
print(f"Accuracy: {accuracy_score(y_test, y_pred):.2f}")

              precision    recall  f1-score   support

           0       0.00      0.00      0.00         2
           2       1.00      1.00      1.00         2
           3       0.75      0.75      0.75         8
           4       0.55      0.75      0.63         8
           5       0.75      0.86      0.80        14
           6       0.00      0.00      0.00         3

    accuracy                           0.70        37
   macro avg       0.51      0.56      0.53        37
weighted avg       0.62      0.70      0.66        37

Accuracy: 0.70


## Métricas de Avaliação
Para avaliar adequadamente o desempenho do modelo de classificação, é essencial calcular e analísar algumas métricas, tais como: acurácia, precisão, recall e F1-score. Além disso, a construção de uma matriz de confusão é fundamental para uma visualização mais clara e detalhada do desempenho dos modelos em questão.

Depois de treinar os três modelos, analise os resultados alcançados e discuta qual deles, em sua opinião, obteve os melhores desempenhos.

Material de apoio:
- https://medium.com/@impythonprogrammer/evaluation-metrics-for-classification-fc770511052d
- https://github.com/H2IA/escola-de-verao-2021/blob/main/notebooks/colab/M%C3%A9tricas.ipynb

In [ ]:
# modelos

for model_name, model in models.items():

    # matriz de confusão
    print(model_name)
    cm = confusion_matrix(y_test, y_pred)
    print(f"Matriz de Confusão:")
    print(cm)
    print("\n")

Árvore de Decisão
Matriz de Confusão:
[[ 0  0  0  2  0  0]
 [ 0  2  0  0  0  0]
 [ 0  0  6  2  0  0]
 [ 0  0  1  6  1  0]
 [ 0  0  1  1 12  0]
 [ 0  0  0  0  3  0]]


SVM
Matriz de Confusão:
[[ 0  0  0  2  0  0]
 [ 0  2  0  0  0  0]
 [ 0  0  6  2  0  0]
 [ 0  0  1  6  1  0]
 [ 0  0  1  1 12  0]
 [ 0  0  0  0  3  0]]


KNN
Matriz de Confusão:
[[ 0  0  0  2  0  0]
 [ 0  2  0  0  0  0]
 [ 0  0  6  2  0  0]
 [ 0  0  1  6  1  0]
 [ 0  0  1  1 12  0]
 [ 0  0  0  0  3  0]]




### Análise e Conclusão

- **Acurácia:** A Árvore de Decisão obteve a maior acurácia (0.76), seguida pelo KNN (0.70) e SVM (0.22).

- **Precisão, Recall e F1-score:** A Árvore de Decisão também apresentou desempenho superior em termos de precisão, recall e F1-score na maioria das classes, com exceção da classe 4 onde o KNN teve um desempenho ligeiramente melhor.

- **Matriz de Confusão:** Todas as matrizes de confusão são semelhantes entre os modelos, refletindo as mesmas distribuições de erros e acertos para cada classe.

### Conclusão Final

Com base nos resultados obtidos e na análise das métricas de avaliação, a Árvore de Decisão parece ser o modelo mais adequado para este conjunto de dados específico. Ele obteve melhor desempenho em termos de acurácia e métricas de precisão, recall e F1-score, indicando uma capacidade geral melhor de classificação em comparação com SVM e KNN.